### DISCUSSION TITLE
How does vcpkg automatically add libraries when I build a project using a port?

### DISCUSSION THREAD

**Discussion by jdx-john** [microsoft/vcpkg#18954] [Q&A]
I'm porting some use of libraries in our solution to use vcpkg, which means clearing out a lot of project-settings for includes, library paths and individual library modules.

I was thinking how amazing vcpkg is to automagically include/link things as needed, is there a clear description somewhere how it actually works (at a high-ish level)? I was particularly interested how .lib files are added in as needed.

**Comment by PhoebeHui** [comment]
@jdx-john, for cmake project, projects configured with the Vcpkg toolchain file will have the appropriate Vcpkg folders added to the cmake search paths. This makes all libraries available to be found through find_package(), find_path(), and find_library().

For msbuild project: 'vcpkg integrate install' this will implicitly add Include Directories, Link Directories, and Link Libraries for all packages installed with Vcpkg to all VS2015, VS2017 and VS2019 MSBuild projects. https://github.com/microsoft/vcpkg/blob/master/scripts/buildsystems/msbuild/applocal.ps1 will help to copy required dll to your project.

See https://github.com/microsoft/vcpkg/blob/master/docs/users/integration.md for details.

**Comment by PhoebeHui** [comment]
@jdx-john, does this still be a question for you?

**Comment by jdx-john** [comment]
I can see how it adds include/library paths, I don't understand how projects are able to automatically find and link .lib files, which need to be explicitly specified at link time? Even using Nuget you have to tell each project its dependencies, this seems so smooth it's like magic :)

**Reply by BillyONeal** [reply]
> I can see how it adds include/library paths, I don't understand how projects are able to automatically find and link .lib files, which need to be explicitly specified at link time? Even using Nuget you have to tell each project its dependencies, this seems so smooth it's like magic :)

The msbuild integration adds literally every lib file in the installed tree to the linker command line.

**Comment by PhoebeHui** [comment]
For cmake project, we have cmake targets file, which defines the libs that need to be linked explicitly when use find_package. but these doesn't provide any native cmake targets file, the libs can't be found automatically, you need to explicitly define the libs via find_library, then it can be found automatically.

For msbuild project, [vcpkg.targets](https://github.com/microsoft/vcpkg/blob/master/scripts/buildsystems/msbuild/vcpkg.targets#L47-L58) will add the libs directories to AdditionalLibraryDirectories, and all libs to AdditionalDependencies. so it can find the libs automatically.

**Comment by PhoebeHui** [comment]
@jdx-john, does that answer your questions?
